apm-sniffer/apm-toolkit-activation/apm-toolkit-log4j-2.x-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/log/log4j/v2/x/async/AsyncAppenderMethodInterceptor.java

```diff
@@ -32,7 +32,9 @@ public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allAr
         MethodInterceptResult result) throws Throwable {
         if (allArguments[0] instanceof EnhancedInstance) {
             EnhancedInstance instances = (EnhancedInstance) allArguments[0];
-            instances.setSkyWalkingDynamicField(ContextManager.getGlobalTraceId());
+            if (instances.getSkyWalkingDynamicField() == null){
+                instances.setSkyWalkingDynamicField(ContextManager.getGlobalTraceId());
```

Comment:
If this is the async mode, it means, it is in another thread. How `ContextManager.getGlobalTraceId()` works?

Reply:
> > But deserialization and then serialization will make Log4jLogEvent lose SkyWalkingDynamicField, which cause no trace id in log

OK, I just read your issue. If this is the case, why don't you enhance `createMemento` to copy this new field?

Reply:
I have the same problem as Sheng? @xuanyu66

Reply:
In our scenario , We remove AsyncAppenderInstrumentation and AsyncLoggerConfigInstrumentation, and the plugin works well.
But if I remove them in open source code, I'am not sure it will break others' code.
So I decide to keep the code, and make it compatible

Reply:
@xuanyu66 We don't discuss under the case, what you did in the private. 
In the community, we all focus on the public and upstream version only. So, forgive me can't give any feedback about your description.
We replied based on your case. You should explain why our ideas are not good or correct.

Reply:
> If this is the case, why don't you enhance createMemento to copy this new 

@wu-sheng But I did  enhance createMemento to copy this new field.

Reply:
What I've done here is to avoid AsyncAppenderMethodIntercepto to cover trace id with null , because as you said, ContextManager.getGlobalTraceId() doesn's works , it's another thread

Reply:
I didn't read the log4j2 source codes as you did. I am confused. You said, there is an issue caused by the object clone, which means you need to propagate the context through copy/paste the dynamic field. But clearly, this code doesn't match your description. This is about setting the current trace id to a new object's dynamic field.

I think we miss some context here. If you feel the words are hard to understand, you could draw somethings(diagram) to show the issue.

Reply:
You can see CreateMementoInstrumentation, which propagates the context through copy/paste the dynamic field.
AsyncAppenderInstrumentation isn't written by me, I just modify it.
> This is about setting the current trace id to a new object's dynamic
the AsyncAppenderInstrumentation and AsyncLoggerConfigInstrumentation is not added by me. I don't why do this

Reply:
So, you mean there is an override that happens here? So you add `==null` to avoid that?

Reply:
@J-Cod3r Are you still around? Could you share what happens here?

Reply:
@EvanLjp Could you take a look at whether this is in use if you activated the async appender?

Reply:
> What I've done here is to avoid AsyncAppenderMethodIntercepto to cover trace id with null , because as you said, ContextManager.getGlobalTraceId() doesn's works , it's another thread

@xuanyu66
This should not be the reason, AsyncAppender will put Log4jLogEvent into the queue, then another thread will take it out and print the log, AsyncAppenderMethodInterceptor intercept before putting it into the queue
![image](https://user-images.githubusercontent.com/7534743/111868712-72eb2b80-89b6-11eb-82d4-bf615b32a6b8.png)

Reply:
@J-Cod3r thanks , I will resubmit

Reply:
@xuanyu66 When you are ready, ask @J-Cod3r to review. He is more qualified than me.